### PR TITLE
[KIALI-357] consolidate state and callbacks in ServiceGraphPage

### PR DESCRIPTION
**Pull request by vnugent**
Graph data, namespace, interval and layout are now in one place, `ServiceGraphPage`.   Browser URL will set namespace, interval and layout accordingly. 
1. Make sure Namespace, Layout, Interval are used consistently.  For example, `CytoscapeLayout.namespace` prop is now of Namespace type.  You can no longer pass a string to it.
2. `CytoscapeLayout`: move graph elements state and refresh logic to its parent.
3. `GraphFilter`: propagates namespace, filter, layout changes to parent via new separate callbacks, `filterChange(newVal)`, `namespaceChange(newVal)` and `layoutChange(newVal)`.  The callback payload contains the new value. `GraphFilters` getters/setters are no longer needed. 
4. Since I introduced `Interval` type and recent changes renamed it to `Duration` in order to keep my rebase/merge simple I decided to override the name change with mine ie keep it as Interval.  I'll address the rename in a separate PR.

Please review!

### CODE REVIEW

**Comment by vnugent** [comment]
kiali 1

**Comment by vnugent** [comment]
kiali-service-graph

File: src/components/CytoscapeLayout/CytoscapeLayout.tsx [hunk 1]
```diff
@@ -63,20 +60,12 @@ export default class CytoscapeLayout extends React.Component<CytoscapeLayoutProp
...
-    if (nextProps.namespace !== this.props.namespace || nextProps.duration !== this.props.duration) {
-      this.updateGraphElements(nextProps);
-    }
   }
 
   shouldComponentUpdate(nextProps: CytoscapeLayoutProps, nextState: CytoscapeLayoutState) {
-    return this.state.loading !== nextState.loading || this.props.layout !== nextProps.layout;
+    // graph data, namespace and layout are now props.
+    // always re-render when those props change.
+    return true;
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply src/components/CytoscapeLayout/CytoscapeLayout.tsx]: +1, this should be changed for clarity.

File: src/components/CytoscapeLayout/__tests__/CytoscapeLayout.test.tsx [hunk 2]
```diff
@@ -1,41 +1,55 @@
 import * as React from 'react';
 import { shallow } from 'enzyme';
-import { Button } from 'patternfly-react';
+// import { Button } from 'patternfly-react';
```
Comment (jshaughn) [inline src/components/CytoscapeLayout/__tests__/CytoscapeLayout.test.tsx]: If this is unused may as well remove it.

File: src/components/CytoscapeLayout/__tests__/CytoscapeLayout.test.tsx [hunk 3]
```diff
@@ -1,41 +1,55 @@
...
-    expect(spyUpdateGraphElements).toHaveBeenCalledTimes(1);
-    const buttonWrapper = wrapper.find(Button);
-    buttonWrapper.simulate('click');
-    expect(spyUpdateGraphElements).toHaveBeenCalledTimes(2);
+    const cytoscapeWrapper = wrapper.find(ReactCytoscape);
+    expect(cytoscapeWrapper.prop('elements').elements.nodes).toEqual(GRAPH_DATA[testNamespace].elements.nodes);
+    expect(cytoscapeWrapper.prop('elements').elements.edges).toEqual(GRAPH_DATA[testNamespace].elements.edges);
   });
+
+  //   it('should refresh data on click', () => {
```
Comment (jshaughn) [inline src/components/CytoscapeLayout/__tests__/CytoscapeLayout.test.tsx]: I'd suggest to either remove this or add a comment as to why we should keep it around.

File: src/components/CytoscapeLayout/graphs/LayoutDictionary.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { BreadthFirstGraph } from './BreadthFirstGraph';
```
Comment (jshaughn) [inline src/components/CytoscapeLayout/graphs/LayoutDictionary.ts]: I like this little consolidation.

File: src/components/GraphFilter/GraphFilter.tsx [hunk 5]
```diff
@@ -1,148 +1,74 @@
...
     this.state = {
-      graphDuration: GraphFilters.getGraphDuration(),
-      graphLayout: GraphFilters.getGraphLayout(),
-      graphNamespace: GraphFilters.getGraphNamespace(),
       availableNamespaces: []
     };
   }
 
   componentDidMount() {
+    // API.GetNamespaces() is also called in Services component.
```
Comment (jshaughn) [inline src/components/GraphFilter/GraphFilter.tsx]: May want to add "TODO"

File: src/components/GraphFilter/GraphFilter.tsx [hunk 6]
```diff
@@ -1,148 +1,74 @@
...
     this.setState({ availableNamespaces: response['data'] });
-  }
+  };
 
-  updateDuration(value: string) {
-    GraphFilters.setGraphDuration(value);
-    this.setState({ graphDuration: GraphFilters.getGraphDuration() });
-    this.props.onFilterChange();
-  }
+  updateInterval = (value: string) => {
```
Comment (jshaughn) [inline src/components/GraphFilter/GraphFilter.tsx]: I know you mentioned this but for consistency you will have to immediately follow up with a PR for change to "Duration" terminology.

File: src/components/GraphFilter/QueryOptions.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  '14400': { step: 480, rateInterval: '1m' }, // 4h - 14400/480=30 buckets which is 1 datapoint per 8m
+  '28800': { step: 960, rateInterval: '1m' }, // 8h - 28800/960=30 buckets which is 1 datapoint per 16m
+  '86400': { step: 2880, rateInterval: '1m' }, // 1d - 86400/2880=30 buckets which is 1 datapoint per 48m
+  '604800': { step: 20160, rateInterval: '1m' }, // 7d - 604800/20160=30 buckets which is 1 datapoint per 5.6h
+  '2592000': { step: 86400, rateInterval: '1m' } // 30d - 2592000/86400=30 buckets which is 1 datapoint per 1d
+};
+
+const DEFAULT = GraphQueryOptionsPerDuration['600'];
+
+const get = (key: string) => {
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply src/components/GraphFilter/QueryOptions.ts]: +1, let's not use 'get'.

File: src/pages/ServiceGraph/ServiceGraphPage.tsx [hunk 8]
```diff
@@ -1,56 +1,106 @@
...
+    if (layoutHasChanged || namespaceHasChanged || intervalHasChanged) {
+      const newParams = {
+        namespace: nextNamespace,
+        graphInterval: nextInterval,
+        graphLayout: nextLayout
+      };
+      this.setState({ params: newParams });
+
+      if (!layoutHasChanged) {
+        // we need to explicitily provide namespace and interval because
```
Comment (jshaughn) [inline src/pages/ServiceGraph/ServiceGraphPage.tsx]: typo: explicitly

File: src/components/CytoscapeLayout/__tests__/CytoscapeLayout.test.tsx (hunk 2 above)
Comment (vnugent) [reply src/components/CytoscapeLayout/__tests__/CytoscapeLayout.test.tsx]: noted

File: src/pages/ServiceGraph/ServiceGraphPage.tsx (hunk 8 above)
Comment (vnugent) [reply src/pages/ServiceGraph/ServiceGraphPage.tsx]: noted

**Pull request merged**